torch/autograd/graph.py

```diff
@@ -261,6 +262,256 @@ def unpack_from_cpu(packed):
+
+class _checkpoint_hook(saved_tensors_hooks):
+    def __init__(self):
+        def pack_hook(x):
+            # Snapshot the state of the current checkpoint stack
+            handle = _Handle()
+            current_frames = tuple(checkpoint_stacks[-1])
+            top_frame = current_frames[-1]
+            # We don't detach here, but we clear saved_tensors upon exiting checkpoint
+            top_frame.saved_tensors[handle] = x
+            idx = top_frame.counter
+            top_frame.counter += 1
+            top_frame.weak_handles.append(weakref.ref(handle))
+            assert len(top_frame.weak_handles) == top_frame.counter
+            return idx, handle, current_frames
+
+        def unpack_hook(saved):
+            idx, handle, frames = saved
+            top_frame = frames[-1]
+
+            if not top_frame.exited:
+                # Backward is called inside checkpoint
+                return top_frame.saved_tensors[handle]
+
+            if not top_frame.is_recomputed:
+                # Loop through the nested checkpoints
+                for i, frame in enumerate(frames):
+                    if frame.exited is False or frame.is_recomputed:
+                        continue
+                    # See NOTE [Nested Checkpoint Input Handling]
+                    if frame.is_nested():
+                        args = frame.get_args_from_parent()
+                    else:
+                        assert frame.maybe_args is not None
+                        args = frame.maybe_args
+
+                    with torch.autograd.enable_grad():
+                        do_recomputation(frame.fn, frame, *args)
+            return top_frame.recomputed[idx]
+
+        super().__init__(pack_hook, unpack_hook)
+
+def get_wrapped_fn(fn):
+    # Capture the current context, so we can replay it
+    def wrapped(*args, **kwargs):
+        return fn(*args, **kwargs)
+    return wrapped
+
+def do_recomputation(fn, target_frame: CheckpointFrame, *args):
+    checkpoint_stacks.append([CheckpointFrame(parent=None, fn=None)])
+    curr_frame = checkpoint_stacks[-1][0]
+    with _checkpoint_hook():
+        _ = fn(*args)  # TODO: why do we need the assign the return?!
+
+    for (idx, weak_handle) in enumerate(curr_frame.weak_handles):
+        if weak_handle() is None:
+            # We may have saved this variable at once point, but it was cleared possibly
+            # because the user called backward inside checkpoint.
+            continue
+        t = curr_frame.saved_tensors[weak_handle()]
+        target_frame.recomputed[idx] = t if idx in target_frame.child_args_idx else t.detach()
+
+    target_frame.is_recomputed = True
+    curr_frame.saved_tensors.clear()
+    curr_frame.exited = True
+    checkpoint_stacks.pop()
+
+def _checkpoint(fn, *args, **kwargs):
+    is_first_stack = len(checkpoint_stacks) == 0
+    if is_first_stack:
+        checkpoint_stacks.append([])
+    curr_stack = checkpoint_stacks[-1]
+    is_nested = len(curr_stack) > 0
```

Comment:
is_nested here means nested in the original execution of the function. This means any checkpoint we do inside recomputation is considered nested. We want to make the same classification so that we also save inputs during recomputation.